pkg/controller/machinedeployment/controller.go

```diff
@@ -15,41 +14,344 @@ See the License for the specific language governing permissions and
 	listers "sigs.k8s.io/cluster-api/pkg/client/listers_generated/cluster/v1alpha1"
+	"sigs.k8s.io/cluster-api/pkg/controller/sharedinformers"
 )
 
+// controllerKind contains the schema.GroupVersionKind for this controller type.
+var controllerKind = v1alpha1.SchemeGroupVersion.WithKind("MachineDeployment")
+
 // +controller:group=cluster,version=v1alpha1,kind=MachineDeployment,resource=machinedeployments
 type MachineDeploymentControllerImpl struct {
 	builders.DefaultControllerFns
 
+	// machineClient a client that knows how to consume Machine resources
+	machineClient clientset.Interface
+
 	// lister indexes properties about MachineDeployment
-	lister listers.MachineDeploymentLister
+	mLister  listers.MachineLister
+	mdLister listers.MachineDeploymentLister
+	msLister listers.MachineSetLister
+
+	mListerSynced  cache.InformerSynced
+	mdListerSynced cache.InformerSynced
+	msListerSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
 }
 
 // Init initializes the controller and is called by the generated code
 // Register watches for additional resource types here.
 func (c *MachineDeploymentControllerImpl) Init(arguments sharedinformers.ControllerInitArguments) {
 	// Use the lister for indexing machinedeployments labels
-	c.lister = arguments.GetSharedInformers().Factory.Cluster().V1alpha1().MachineDeployments().Lister()
+	mInformer := arguments.GetSharedInformers().Factory.Cluster().V1alpha1().Machines()
+	msInformer := arguments.GetSharedInformers().Factory.Cluster().V1alpha1().MachineSets()
+	mdInformer := arguments.GetSharedInformers().Factory.Cluster().V1alpha1().MachineDeployments()
+	c.mLister = mInformer.Lister()
+	c.msLister = msInformer.Lister()
+	c.mdLister = mdInformer.Lister()
+	c.mListerSynced = mInformer.Informer().HasSynced
+	c.msListerSynced = msInformer.Informer().HasSynced
+	c.mdListerSynced = mdInformer.Informer().HasSynced
+
+	msInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addMachineSet,
+		UpdateFunc: c.updateMachineSet,
+		DeleteFunc: c.deleteMachineSet,
+	})
+
+	mc, err := clientset.NewForConfig(arguments.GetRestConfig())
+	if err != nil {
+		glog.Fatalf("error building clientset for machineClient: %v", err)
+	}
+	c.machineClient = mc
+
+	c.queue = arguments.GetSharedInformers().WorkerQueues["MachineDeployment"].Queue
 }
 
-// Reconcile handles enqueued messages
+func (c *MachineDeploymentControllerImpl) getMachineSetsForDeployment(d *v1alpha1.MachineDeployment) ([]*v1alpha1.MachineSet, error) {
+	// List all MachineSets to find those we own but that no longer match our
+	// selector. They will be orphaned by ClaimMachineSets().
+	msList, err := c.msLister.MachineSets(d.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: flush out machine set adoption.
+
+	var filteredMS []*v1alpha1.MachineSet
+	for _, ms := range msList {
+		if metav1.GetControllerOf(ms) != nil && !metav1.IsControlledBy(ms, d) {
```

Comment:
If a MS doesn't have a controller, we want to ignore it, I think?  So `if metav1.GetController(ms) == nil || ...`

Reply:
nice catch, was originally thinking of handling it when I added the adoption code, but it makes sense to ignore it for now.
fixed.